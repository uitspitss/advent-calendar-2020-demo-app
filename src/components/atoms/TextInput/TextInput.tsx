import React, { FC, ComponentPropsWithRef } from 'react';
import tw, { styled } from 'twin.macro';
import { UseFormMethods, FieldError } from 'react-hook-form';

export type TextInputProps = {
  children?: never;
  className?: string;
  /**
   * プレースホルダ
   */
  placeholder?: string;
  /**
   * テキストエリアにするかどうか
   */
  textarea?: boolean;
  /**
   * RHF の register
   */
  register?: UseFormMethods['register'];
  /**
   * RHF の error message
   */
  error?: FieldError;
} & ComponentPropsWithRef<'input'> &
  ComponentPropsWithRef<'textarea'>;

type Props = TextInputProps;

const Component: FC<Props> = (props) => {
  const {
    className,
    placeholder,
    type,
    textarea,
    name,
    register,
    error,
  } = props;

  return (
    <div className={className}>
      <label htmlFor={name}>{name}</label>
      {textarea ? (
        <textarea
          {...{ className, name }}
          id={name}
          placeholder={placeholder || ''}
          aria-invalid={error ? 'true' : 'false'}
          ref={register}
        />
      ) : (
        <input
          {...{ className, name }}
          id={name}
          type={type || 'text'}
          placeholder={placeholder || ''}
          aria-invalid={error ? 'true' : 'false'}
          ref={register}
        />
      )}
      {error && (
        <span className="error" role="alert">
          {error.message}
        </span>
      )}
    </div>
  );
};

const StyledComponent = styled(Component)<Pick<Props, 'textarea'>>`
  & > {
    input,
    textarea {
      ${tw`w-full bg-gray-100 rounded border border-gray-400 focus:outline-none focus:border-indigo-500`}
      ${({ textarea }) => textarea && tw`h-48 resize-none block`}
    }

    .error {
      ${tw`text-red-500`}
    }
  }
`;

/**
 * 汎用的なテキストインプット
 */
export const TextInput: FC<TextInputProps> = (props) => (
  <StyledComponent {...props} />
);
