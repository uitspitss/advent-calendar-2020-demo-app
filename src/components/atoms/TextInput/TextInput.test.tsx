import React from 'react';
import renderer from 'react-test-renderer';

import { TextInput } from './TextInput';

describe('TextInput', () => {
  test('Snap Shot', () => {
    const tree = renderer.create(<TextInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
