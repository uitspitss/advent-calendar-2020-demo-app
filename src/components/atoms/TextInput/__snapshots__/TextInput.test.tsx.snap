// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Snap Shot 1`] = `
.emotion-0 > input,
.emotion-0 > textarea {
  width: 100%;
  --bg-opacity: 1;
  background-color: rgba(247,250,252,var(--bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --border-opacity: 1;
  border-color: rgba(203,213,224,var(--border-opacity));
}

.emotion-0 > input:focus,
.emotion-0 > textarea:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --border-opacity: 1;
  border-color: rgba(102,126,234,var(--border-opacity));
}

.emotion-0 > .error {
  --text-opacity: 1;
  color: rgba(245,101,101,var(--text-opacity));
}

<div
  className="emotion-0 emotion-1"
>
  <label />
  <input
    aria-invalid="false"
    className="emotion-0 emotion-1"
    placeholder=""
    type="text"
  />
</div>
`;
